service/src/test/java/bio/terra/workspace/service/iam/ControlledResourceSamPolicyBuilderTest.java

```diff
@@ -0,0 +1,180 @@
+    var policyBuilder =
+        new ControlledResourceSamPolicyBuilder(
+            ControlledResourceIamRole.EDITOR,
+            null,
+            ControlledResourceCategory.APPLICATION_SHARED,
+            app);
+    var request = new CreateResourceRequestV2();
+
+    assertThrows(
+        InternalLogicException.class,
+        () -> policyBuilder.addPolicies(request),
+        "Specifying a private IAM role on an application shared resource is invalid");
+  }
+
+  @Test
+  void addPolicies_applicationSharedWithNoAppFails() {
+    // ensure an application shared resource cannot have a user with a private IAM role on it
+    // (tho choice of EDITOR for the private IAM role is arbitrary--any value will cause an
+    // exception.
+    var policyBuilder =
+        new ControlledResourceSamPolicyBuilder(
+            ControlledResourceIamRole.EDITOR,
+            null,
+            ControlledResourceCategory.APPLICATION_SHARED,
+            null);
+    var request = new CreateResourceRequestV2();
+
+    assertThrows(InternalLogicException.class, () -> policyBuilder.addPolicies(request));
+  }
+
+  @Test
+  void addPolicies_applicationSharedWithUserEmailFails() {
+    var app =
+        new WsmWorkspaceApplication()
+            .application(new WsmApplication().serviceAccount("fake-svc@example.com"));
+    var policyBuilder =
+        new ControlledResourceSamPolicyBuilder(
+            ControlledResourceIamRole.EDITOR,
+            "fake@example.com",
+            ControlledResourceCategory.APPLICATION_SHARED,
+            app);
+    var request = new CreateResourceRequestV2();
+
+    assertThrows(InternalLogicException.class, () -> policyBuilder.addPolicies(request));
+  }
+
+  @Test
+  void addPolicies_applicationPrivate() {
+    var policyBuilder =
+        new ControlledResourceSamPolicyBuilder(
+            ControlledResourceIamRole.WRITER,
+            "fake@example.com",
+            ControlledResourceCategory.APPLICATION_PRIVATE,
+            new WsmWorkspaceApplication()
+                .application(new WsmApplication().serviceAccount("fake-svc@example.com")));
+    var request = new CreateResourceRequestV2();
+
+    policyBuilder.addPolicies(request);
+
+    assertThat(
+        request.getPolicies().get(ControlledResourceIamRole.EDITOR.toSamRole()).getMemberEmails(),
+        contains("fake-svc@example.com"));
+    assertThat(
+        request.getPolicies().get(ControlledResourceIamRole.WRITER.toSamRole()).getMemberEmails(),
+        contains("fake@example.com"));
+  }
+
+  @Test
+  void addPolicies_applicationPrivateWithNoAppFails() {
+    var policyBuilder =
+        new ControlledResourceSamPolicyBuilder(
+            ControlledResourceIamRole.EDITOR,
+            null,
+            ControlledResourceCategory.APPLICATION_PRIVATE,
+            null);
+    var request = new CreateResourceRequestV2();
+
+    assertThrows(InternalLogicException.class, () -> policyBuilder.addPolicies(request));
```

Comment:
Since an InternalLogicException can be thrown for a couple of different reasons, it would be good to check the error message.

And is it also a problem in this case that this is APPLICATION_PRIVATE, but has no email address for the EDITOR user?

Reply:
Updated to check the error, and passed in an email, good idea -- thanks!